serving/serving.go

```diff
@@ -0,0 +1,201 @@
+//
+// It supports listening on specific ports or randomly-available ports, and
+// reports the corresponding bind addresses. This is useful for creating
+// multiple servers in tests where ports may conflict. The server also
+// gracefully stops requests when the provided context is closed.
+package serving
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"net/http"
+	"strconv"
+	"time"
+
+	"github.com/abcxyz/pkg/logging"
+	"google.golang.org/grpc"
+)
+
+// Server provides a gracefully-stoppable http server implementation.
+type Server struct {
+	ip       string
+	port     string
+	listener net.Listener
+}
+
+// New creates a new HTTP server listening on the provided address that responds
+// to the http.Handler. It starts the listener, but does not start the server.
+// If an empty port is given (or port "0"), the server randomly chooses one.
+//
+// It binds to all interfaces (0.0.0.0, [::]). To bind to specific interfaces,
+// use [NewFromListener] instead.
+func New(port string) (*Server, error) {
+	// Create the net listener first, so the connection ready when we return. This
+	// guarantees that it can accept requests.
+	addr := fmt.Sprintf(":" + port)
+	listener, err := net.Listen("tcp", addr)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create listener on %s: %w", addr, err)
+	}
+
+	return NewFromListener(listener)
+}
+
+// NewFromListener creates a new server on the given listener. This is useful if
+// you want to customize the listener type or bind custom networks more than
+// [New] allows.
+func NewFromListener(listener net.Listener) (*Server, error) {
+	netAddr := listener.Addr()
+	addr, ok := netAddr.(*net.TCPAddr)
+	if !ok {
+		return nil, fmt.Errorf("listener is not tcp (got %T)", netAddr)
+	}
+
+	return &Server{
+		ip:       addr.IP.String(),
+		port:     strconv.Itoa(addr.Port),
+		listener: listener,
+	}, nil
+}
+
+// Addr returns the server's listening address (ip + port).
+func (s *Server) Addr() string {
+	return net.JoinHostPort(s.ip, s.port)
+}
+
+// IP returns the server's listening IP.
+func (s *Server) IP() string {
+	return s.ip
+}
+
+// Port returns the server's listening port.
+func (s *Server) Port() string {
+	return s.port
+}
+
+// StartHTTP starts the given [net/http.Server] and blocks until the provided
+// context is closed. When the provided context is closed, the HTTP server is
+// gracefully stopped with a timeout of 10 seconds; once a server has been
+// stopped, it is NOT safe for reuse.
+//
+// Note that the incoming [net/http.Server]'s address is ignored over the
+// listener's configuration.
+func (s *Server) StartHTTP(ctx context.Context, srv *http.Server) error {
+	logger := logging.FromContext(ctx)
+
+	// Start the server in a background goroutine so we can listen for cancellation
+	// in the main process.
+	errCh := make(chan error, 1)
+	doneCh := make(chan struct{})
+	go func() {
+		defer close(doneCh)
+
+		logger.Infow("server is starting", "ip", s.ip, "port", s.port)
```

Comment:
I'm curious, why use the non-sentence convention for log messages? Wouldn't it make sense for these?